src/prometheus/parsers.rs

```diff
@@ -881,16 +881,51 @@ pub fn parse_prometheus(
             Vec::new()
         };
 
-        let (label_names, label_values) = {
-            let mut names = Vec::new();
-            let mut values = Vec::new();
-            for (name, value) in labels.into_iter() {
-                names.push(name.to_owned());
-                values.push(value.to_owned());
+        let mut label_names = Vec::new();
+        let mut label_values = Vec::new();
+
+        // Allow samples to have different label sets in non strict mode.
+        if let Some(family_label_names) = family.label_names.as_mut().filter(|_| !strict_mode) {
+            for (name, _) in &labels {
+                let label_name = name.to_string();
+                let reserved_label_name = (family.family_type == Some(PrometheusType::Histogram) && label_name == "le") ||
+                                          (family.family_type == Some(PrometheusType::Summary) && label_name == "quantile");
+
+                // Append any new label names that we haven't seen yet to this metric family's overall list of label names. Skip reserved label names.
+                if !family_label_names.names.contains(&label_name) && !reserved_label_name {
+                    family_label_names.names.push(label_name.clone());
+
+                    // Update all family metrics parsed so far (excluding the current sample) to append an empty string for the missing label value.
+                    for metric in family.metrics.iter_mut() {
+                        metric.label_values.push("".to_string());
+                    }
+                }
             }
 
-            (names, values)
-        };
+            // Create a map for faster label name lookups in the current sample being parsed.
+            let labels_map: std::collections::HashMap<_, _> = labels.into_iter().collect();
+
+            // Create the label names and values vectors for the current sample, following the order in `family.label_names`. Use empty string label values for any labels not present in this sample.
+            for label_name in family_label_names.names.iter() {
```

Comment:
More concise: 

```suggestion
            for label_name in &family_label_names.names {
```

This library doesn't have it enabled, but I turned on clippy temporarily for this module by adding this in my editor:

```diff
+ #[warn(clippy::pedantic)]
+ #[allow(clippy::too_many_lines)]
mod parsers;
```

Reply:
Ah, didn't realize it was off. Have refactored generally after turning it on see last commit.